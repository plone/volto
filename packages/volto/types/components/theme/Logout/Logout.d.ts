export default Logout;
declare function Logout({ location }: {
    location: any;
}): JSX.Element;
