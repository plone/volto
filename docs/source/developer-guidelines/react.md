---
myst:
  html_meta:
    "description": "TBD"
    "property=og:description": "TBD"
    "property=og:title": "React"
    "keywords": "Volto, Plone, frontend, React"
---

# React

```{note}
This documentation is a work in progress. Any help is welcome to fill in the
gaps!
```
