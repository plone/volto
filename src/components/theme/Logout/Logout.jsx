import { useEffect, useMemo } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { defineMessages, useIntl } from 'react-intl';
import qs from 'query-string';
import { Login, Toast } from '@plone/volto/components';
import { logout, purgeMessages } from '@plone/volto/actions';
import { toast } from 'react-toastify';

const messages = defineMessages({
  loggedOut: {
    id: 'Logged out',
    defaultMessage: 'Logged out',
  },
  loggedOutContent: {
    id: 'You have been logged out from the site.',
    defaultMessage: 'You have been logged out from the site.',
  },
});

const Logout = ({ location }) => {
  const token = useSelector((state) => state.userSession.token, shallowEqual);
  const history = useHistory();
  const dispatch = useDispatch();
  const intl = useIntl();

  const returnUrl = useMemo(
    () =>
      qs.parse(location.search).return_url ||
      location.pathname
        .replace(/\/login\/?$/, '')
        .replace(/\/logout\/?$/, '') ||
      '/',
    [location],
  );

  useEffect(() => {
    dispatch(logout());
    dispatch(purgeMessages());
  }, [dispatch]);

  useEffect(() => {
    if (!token) {
      history.replace(returnUrl || '/');
      if (!toast.isActive('loggedOut')) {
        toast.info(
          <Toast
            info
            title={intl.formatMessage(messages.loggedOut)}
            content={intl.formatMessage(messages.loggedOutContent)}
          />,
          { autoClose: false, toastId: 'loggedOut' },
        );
      }
    }
  }, [history, returnUrl, intl, token]);

  return <Login location={{ query: location.query }} isLogout={true} />;
};

export default Logout;
