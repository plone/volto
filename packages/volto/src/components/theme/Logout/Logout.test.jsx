import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';

import Logout from './Logout';

const mockStore = configureStore();

jest.mock('../Login/Login', () => jest.fn(() => <div />));

describe('Logout', () => {
  it('renders a logout component', () => {
    const store = mockStore({
      userSession: {
        login: {},
      },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <MemoryRouter>
            <Logout location={{ pathname: '' }} />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
