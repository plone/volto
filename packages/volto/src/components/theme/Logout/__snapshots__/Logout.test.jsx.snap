// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout renders a logout component 1`] = `<div />`;
